[FLINK-1525]Introduction of a small input parameter parsing utility

rmetzger opened the pull request:
Hi,
last week I was running a bunch of Flink Streaming jobs on a cluster. One of the jobs had 8 arguments which I changed in different iterations of the program.
I ended up passing arguments like
```
16 1 8 3 10k hdp22-w-1.c.internal:6667,hdp22-w-0.c.internal:6667,hdp22-m.c.internal:6667 10000
```
Its obvious that this is not easily maintainable.
In addition to this experience, I got similar feedback from at least two other Flink users.

Therefore, I sat down and implemented a simple class which allows users to work with input parameters in a hassle-free manner.
The tool is called **ParameterUtil**. It can be initialized from:
- regular command line arguments (`-` and `--`): `ParameterUtil.fromArgs(new String[]{"--berlin"});`
- `.properties` files: `ParameterUtil.fromPropertiesFile(propertiesFile);`
- system properties (-D arguments to the JVM): `ParameterUtil.fromSystemProperties()`;

I'm also planning to provide an initializer which accepts the same arguments as Hadoop's GenericOptionsParser: https://hadoop.apache.org/docs/r1.0.4/api/org/apache/hadoop/util/GenericOptionsParser.html (our users are just too used to Hadoop's tooling)

For accessing arguments, it has methods like:
`parameter.getRequired("input")`, `parameter.get("output", "myDefaultValue")`, `parameter.getLong("expectedCount", -1L)` and so on ...

Also, I added a method to export the parameters to Flink's `Configuration` class:
```
Configuration config = parameter.getConfiguration();
config.getLong("expectedCount", -1L)
```
This allows users to pass the input arguments to operators in the APIs:
```
text.flatMap(new Tokenizer()).withParameters(conf)
```

The `ParameterUtil` itself is Serializable, so it can be passed into user functions (for example to the `Tokenizer`).
Also, I extended the `ExecutionConfig` to allow passing a `UserConfig` with custom stuff inside it.

The `ParameterUtil` is implementing the `UserConfig` interface, so users can do the following:

```java
public static void main(String[] args) throws Exception {
  ParameterUtil pt = ParameterUtil.fromArgs(args);
  final ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
  env.getConfig().setUserConfig(pt);
  ///.... regular flink stuff ....
}
```
Inside a (rich) user function, users can access the command line arguments:
```java
text.flatMap(new Tokenizer()).flatMap(new RichFlatMapFunction<Tuple2<String, Integer>, Tuple2<String, Integer>>() {
	@Override
	public void flatMap(Tuple2<String, Integer> value, Collector<Tuple2<String, Integer>> out) throws Exception {
		ExecutionConfig.UserConfig uc = getRuntimeContext().getExecutionConfig().getUserConfig();
		ParameterUtil pt = (ParameterUtil) uc;
		float norm = pt.getFloat("normalization", 0.15f);
	}
})
```

The `UserConfig` allows to export Key/Value pairs to the web interface. Running Wordcount:
```
/bin/flink run ./examples/flink-java-examples-0.9-SNAPSHOT-WordCount.jar --input /home/robert/incubator-flink/build-target/README.txt --output /tmp/wo
```
Will lead to the following result:

![paramutil](https://cloud.githubusercontent.com/assets/89049/7550566/14ea36c2-f667-11e4-9a81-ee6a017527b0.png)


Before I'm now going to add this to all examples I would like to get some feedback for the API choices I made (I don't want to change all examples afterwards ;) ).
Wordcount currently looks like this:
```java
public static void main(String[] args) throws Exception {
	ParameterUtil pt = ParameterUtil.fromArgs(args);
	boolean fileOutput = pt.getNumberOfParameters() == 2;
	String textPath = null;
	String outputPath = null;
	if(fileOutput) {
		textPath = pt.getRequired("input");
		outputPath = pt.getRequired("output");
	}
	
	// set up the execution environment
	final ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
	env.getConfig().setUserConfig(pt);
	
	// create initial DataSet, containing the text lines.
	DataSet<String> text;
	if(fileOutput) {
		text = env.readTextFile(textPath);
	} else {
		// get default test text data
		text = WordCountData.getDefaultTextLineDataSet(env);
	}
	
	DataSet<Tuple2<String, Integer>> counts =
		// split up the lines in pairs (2-tuples) containing: (word,1)
		text.flatMap(new Tokenizer())
		// group by the tuple field "0" and sum up tuple field "1"
		.groupBy(0)
		.sum(1);
```

uce commented:
**Feedback w/o code review**

I like the idea very much. +1 for the general thing. Exporting it to the web interface is great. Thanks for this. :-)

The Hadoop compatibility idea is also very nice. Let's make sure to follow up with it soon (file a JIRA etc.)

Regarding adding this to all tests: I think we should discuss this. I am undecided. On the one hand, it will break everything that shows how to run the examples, because you will have to specify the parameter names now. On the other hand, I like that people will be exposed to the utility. I think I'm leaning towards option 2 atm.

What's currently missing is documentation.

---

In the word count example: is the only point of setting the user config in the execution config only to show the values in the web interface?

tillrohrmann commented:
Cool tool, Robert. This is definitely useful for users.

If this tool also supported positional arguments, then we wouldn't have to
change how the examples are called. Furthermore, it wouldn't force you to
name all of your parameters if you don't want to.
On May 10, 2015 10:29 AM, "Ufuk Celebi" <notifications@github.com> wrote:

> *Feedback w/o code review*
>
> I like the idea very much. +1 for the general thing. Exporting it to the
> web interface is great. Thanks for this. :-)
>
> The Hadoop compatibility idea is also very nice. Let's make sure to follow
> up with it soon (file a JIRA etc.)
>
> Regarding adding this to all tests: I think we should discuss this. I am
> undecided. On the one hand, it will break everything that shows how to run
> the examples, because you will have to specify the parameter names now. On
> the other hand, I like that people will be exposed to the utility. I think
> I'm leaning towards option 2 atm.
>
> What's currently missing is documentation.
> ------------------------------
>
> In the word count example: is the only point of setting the user config in
> the execution config only to show the values in the web interface?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/flink/pull/664#issuecomment-100600902>.
>

rmetzger commented:
Thank you for the feedback so far!

@uce 
> In the word count example: is the only point of setting the user config in the execution config only to show the values in the web interface?

Yes. But I think its worth it, because it'll teach our users how to use all features of the util.

>If this tool also supported positional arguments ...

I also though about adding those, but decided against it, because args[n] is already a way of accessing the arguments by their position ;)
But I'll add it so that users can also specify default values .. and we take care of the parsing.

hsaputra commented:
This is great idea @rmetzger 

I am bit confuse about the addition of UserConfig interface. 
Is the interface added to add abstraction for the ParameterUtil? It is a bit confusing what does user mean? is it the program driver or actual user (ie the principal) which executing the Flink program?

rmetzger commented:
In Hadoop, the `UserConfig` would probably be a `Configuration` object with key/value pairs.
In Flink, we are trying to get rid of these untyped maps.
Instead, I would recommend users to use a simple java class, like
```java
public static class MyConfig extends UserConfig {
  public long someLongValue;
  public int someInt;
  // this is optional
  public Map<String, String> toMap() { return null; }
}
```
It can be used in a similar way to a Configuration object, but the compiler is able to check the types.

The `ParameterUtil` is implementing the UserConfig interface to expose the configuration values through the Flink program & in the web interface.

aljoscha commented:
I like it. But I think it needs some functionality for verifying parameters. To let the user specify some parameters that always need to be there and a description of the parameter. Similar to how other tools print the "usage" when you don't give correct arguments.

rmetzger commented:
I agree, however it should be optional. I don't like these tools where you spend a lot of time registering / specifying arguments.
People want to analyze their data, not configure a huge parameter parsing framework ;)
But I'm going to look into this and see how much work it would be to implement it. If it blocks me from getting this PR merged soon, I'll file a follow-up JIRA.

mxm commented:
Very helpful utility. I think it's worth adapting all the examples if we merge this. Removes a lot of unnecessary code and makes the examples more readable. 

May I ask why you didn't reuse the Parser in `org.apache.commons.cli`? Too much overhead?

rmetzger commented:
Thank you.
I didn't reuse the parser there because there is an ongoing JIRA to unify all the command line parsers inside Flink. It seems that we are using at least two different libraries .. and the consensus in the JIRA seems to be using a third one to solve the problem.
But I didn't want to add more confusion to the topic than we already have.

mxm commented:
Thanks for clarifying. I agree that the parser in Apache Commons is not the nicest...

rmetzger commented:
>>If this tool also supported positional arguments ...
>
>I also though about adding those, but decided against it, because args[n] is already a way of > accessing the arguments by their position ;)
> But I'll add it so that users can also specify default values .. and we take care of the parsing.

I'm not so sure about this anymore. Positional arguments would break a lot in the design of the  `ParameterUtil`.
For example the export to the web interface, the Configuration object or Properties.

rmetzger commented:
@aljoscha:

Is this the API you were thinking of?
```java
RequiredParameters required = new RequiredParameters();
Option input = required.add("input").alt("i").description("Path to input file or directory"); // parameter with long and short variant
required.add("output"); // parameter only with long variant
Option parallelism = required.add("parallelism").alt("p").type(Integer.class); // parameter with type
Option spOption = required.add("sourceParallelism").alt("sp").defaultValue(12).description("Number specifying the number of parallel data source instances"); // parameter with default value, specifying the type.

ParameterUtil parameter = ParameterUtil.fromArgs(new String[]{"-i", "someinput", "--output", "someout", "-p", "15"});

required.check(parameter);
required.printHelp();
required.checkAndPopulate(parameter);

String inputString = input.get();
int par = parallelism.getInteger();
String output = parameter.get("output");
int sourcePar = parameter.getInteger(spOption.getName());
```

aljoscha commented:
Yes, something like this.

hsaputra commented on flink-examples/flink-java-examples/src/main/java/org/apache/flink/examples/java/wordcount/PojoExample.java:
```diff
@@ -26,6 +26,7 @@
 import org.apache.flink.util.Collector;
 
 
+
```
Extra new line?

hsaputra commented:
Thanks for clarifying, Robert.

I think convention to expect *Util classes to just contain static methods instead of being created as an instance.
Maybe we could use ```InputParameters``` as class name to help parse the CLI arguments instead?

StephanEwen commented:
Nice idea. Do we need the special `UserConfig` interface, or can we use a `Properties` object, or a directly a `Map<String, String>`?

rmetzger commented:
I thought its much nicer for users to provide a typed POJO (implementing `UserConfig`) instead of a String map.
This approach leaves at least the door open for a typed configuration object. If you want, 
I'll also let the `Configuration` class implement the UserConfig interface, so that users can used untyped key/value configuration there as well.

StephanEwen commented:
Does every user need to implement the user config object?
In the code, do you always need to cast the user config to your specific class?

rmetzger commented:
Yes. But my PR will provide two standard implementations (ParameterTool, Configuration), so only if users want more features they need to implement the config.
Users have to cast it in their function, yes.

@hsaputra: I renamed the class to `ParameterTool`.

----------------------------------

I think the pull request is in a mergeable state now. I didn't change the examples yet. Thats going to be a big change which I would like to do separately, once this is merged.

Please review the PR ...

mxm commented:
Do you want to allow empty keys like in `-- value` or even `--`? I think the parser should throw an exception in this case.

rmetzger commented:
Good catch, thank you.
I added test cases that validate we're throwing an exception for `-` and `--` inputs.

hsaputra commented on flink-java/src/test/java/org/apache/flink/api/java/utils/ParameterToolTest.java:
```diff
@@ -0,0 +1,220 @@
+
+
+
+	/*@Test
```
Do you want to keep this test as optional? If yes then it is better to have comments on why and when to uncomment this test. Otherwise, let's just remove it for now.

rmetzger replied:
I'll remove it

hsaputra commented:
Hi @rmetzger, just did a pass and other than comments about unused test and broken Travis due to check style I think this PR is ready to go. +1

uce commented on flink-runtime/src/main/resources/log4j.properties:
```diff
@@ -18,7 +18,7 @@
 
 # Convenience file for local debugging of the JobManager/TaskManager.
-log4j.rootLogger=OFF, console
+log4j.rootLogger=INFO, console
```
back to OFF?

rmetzger replied:
I'll very soon fix that issue so that we don't have to change the logging level every time we want to debug something.

uce commented on docs/apis/best_practices.md:
```diff
@@ -0,0 +1,155 @@
+Almost all Flink applications, both batch and streaming rely on external configuration parameters.
+For example for specifying input and output sources (like paths or addresses), also system parameters (parallelism, runtime configuration) and application specific parameters (often used within the user functions).
+
+Since version 0.9 we are providing a simple called `ParameterTool` to provide at least some basic tooling for solving these problems.
```
a simple ... => utility?

uce commented on docs/apis/best_practices.md:
```diff
@@ -0,0 +1,155 @@
+
+{% highlight java %}
+ParameterTool parameters = ParameterTool.fromArgs(args);
+DataSet<Tuple2<String, Integer>> counts = text.flatMap(new Tokenizer()).setParallelism(parameters.getInt("mapParallelism", 2));
```
Maybe to make this patter readable... do
```
int parallelism = parameters.get("mapParallelism", 2);
counts = ....setParalellism(parallelism);
```

uce commented on docs/apis/programming_guide.md:
```diff
@@ -665,6 +667,18 @@ DataSet<Integer> result = in.partitionByHash(0)
     </tr>
     <tr>
+      <td><strong>Custom Partitioning</strong></td>
+      <td>
```
This is unrelated to the change, right?

rmetzger replied:
Yes, its a fix to https://issues.apache.org/jira/browse/FLINK-1260 ;)

uce commented on docs/apis/best_practices.md:
```diff
@@ -0,0 +1,155 @@
+
+Since version 0.9 we are providing a simple called `ParameterTool` to provide at least some basic tooling for solving these problems.
+
+As you'll see Flink is very flexible when it comes to parsing input parameters. You are free to choose any other framework, like [Commons CLI](https://commons.apache.org/proper/commons-cli/), [argparse4j](http://argparse4j.sourceforge.net/), or others.
```
I think this paragraph is confusing. I would remove it

rmetzger replied:
I rephrased it

uce commented:
Thanks for the docs update! Very nice idea with the best practices page. +1

rmetzger commented:
Thanks for the review Henry and Ufuk!
I addressed your concerns and updated the PR.

I rebased the PR to the current master and filed two JIRAs for missing features
https://issues.apache.org/jira/browse/FLINK-2018
https://issues.apache.org/jira/browse/FLINK-2017

I'll probably merge this change in the next 12 hours ...

aljoscha commented:
+1

rmetzger commented:
Okay .. merging
